go-controller/pkg/level_driven_controller/level_driven_controller.go

```diff
@@ -0,0 +1,271 @@
+	c.queue.ShutDown()
+	if c.eventHandler != nil {
+		if err := c.informer.RemoveEventHandler(c.eventHandler); err != nil {
+			klog.Errorf("Failed to remove event handler for controller %s: %w", c.name, err)
+		}
+	}
+}
+
+func (c *controller[T]) onAdd(objInterface interface{}) {
+	newObj, ok := objInterface.(*T)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("controller %s: expecting %T but received %T", c.name, *new(T), newObj))
+		return
+	}
+	if !c.handler.ObjNeedsUpdate(nil, newObj) {
+		return
+	}
+
+	key, err := cache.MetaNamespaceKeyFunc(newObj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("controller %s: couldn't get key for object %+v: %v", c.name, newObj, err))
+		return
+	}
+	c.queue.Add(key)
+}
+
+func (c *controller[T]) onUpdate(oldObjInterface, newObjInterface interface{}) {
+	oldObj, ok := oldObjInterface.(*T)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("controller %s: expecting %T but received %T", c.name, *new(T), oldObj))
+		return
+	}
+	newObj, ok := newObjInterface.(*T)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("controller %s: expecting %T but received %T", c.name, *new(T), newObj))
+		return
+	}
+
+	if !c.handler.ObjNeedsUpdate(oldObj, newObj) {
+		return
+	}
+
+	key, err := cache.MetaNamespaceKeyFunc(newObj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("controller %s: couldn't get key for object %+v: %v", c.name, newObj, err))
+		return
+	}
+	c.queue.Add(key)
+}
+
+func (c *controller[T]) onDelete(objInterface interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(objInterface)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("controller %s: couldn't get key for object %+v: %v", c.name, objInterface, err))
+		return
+	}
+	c.queue.Add(key)
+}
+
+// processNextQueueItem returns false when the queue is shutdown and the handling should be stopped.
+// Otherwise, it handles the next item from the queue and always returns true.
+func (c *controller[T]) processNextQueueItem() bool {
+	key, shutdown := c.queue.Get()
+
+	if shutdown {
+		return false
+	}
+
+	defer c.queue.Done(key)
+
+	err := c.handler.Reconcile(key.(string))
+	if err != nil {
+		if c.queue.NumRequeues(key) < maxRetries {
+			klog.Infof("Controller %s: error found while processing %s: %v", c.name, key.(string), err)
+			c.queue.AddRateLimited(key)
+			return true
+		}
+		klog.Warningf("Controller %s: dropping %s out of the queue: %v", c.name, key.(string), err)
+		utilruntime.HandleError(err)
+	}
+	c.queue.Forget(key)
+	return true
+}
+
+func (c *controller[T]) AddToQueue(key string) {
+	c.queue.Add(key)
+}
+
+// StoppableController is a level-driven controller that is shut down after Stop() call.
+// It creates its own stop channel and waiting group to ensure proper cleanup.
+type StoppableController[T any] struct {
+	controller[T]
+	localStopChan chan struct{}
+	wg            *sync.WaitGroup
+}
+
+func NewStoppableController[T any](name string, config *ControllerConfig[T]) StoppableController[T] {
```

Comment:
since I have 2 types: `StoppableController` and `StopChanController` I am not sure what is the best naming here.
We could say e.g. that `StopChanController` is the default one, and call it Controller, with New method

Reply:
The fact that you have two is in itself confusing. We should just have one and users should deal with it.